// stem-daemon/lib/plugins/disk.js
"usr strict";

var JDisk = require('jsdx-disk');
var fs = require('fs');
var async = require('async');

var disk = new JDisk();

var Disk = function() {
	var self = this;

	self.sataDevices = [];
};

Disk.prototype.constructor = function(callback) {
	var self = this;

	disk.init(function() {

		// Initializing current devices
		disk.getDrives(function(err, drives) {

			for (var index in drives) {

				var drive = drives[index];

				// We only handle ATA devices
				if (drive.connectionBus != 'ata')
					continue;

				// Getting all block devices, then mount it
				drive.getBlockDevices(function(err, devices) {

					devices.forEach(function(device, index, arr) {

						// Mounted already
						if (device.mountPoints) {

							self.sataDevices.push({
								deviceName: device.deviceName,
								mountPoints: device.mountPoints
							});
							//console.log(device.deviceName + 'Mounted on ' + device.mountPoints);
							if( device.deviceName == "sda1" && device.mountPoints == ""){
								device.mount({}, function(err, mountPath) {
									if(err)
										{
										//console.log("err1");
										//console.log(err);
										}
									console.log(mountPath);
									async.waterfall([
										function(next) {
											fs.exists('/mnt/share/nas', function (exists) {
												//console.log("function 1");												
												next(null, exists);
											});
										}, function(exists, next) {
											if (exists) {
												fs.unlink('/mnt/share/nas', function(err) {
													//console.log("function 2_1");	
													next();
												});
												return;
												//console.log("function 2_2");	
											}
											//console.log("function 2_3");											
										
											next();
										}, function(next) {
											fs.symlink(mountPath, '/mnt/share/nas', function (err) {
												if (err){
													//console.log("err2");
													//console.log(err);
												} 
												//console.log("function 3");
												fs.chown(mountPath, 1001, 1001);
												fs.chmod(mountPath, '777');
												fs.chown('/mnt/share/nas', 1001, 1001);
												fs.chmod('/mnt/share/nas', '777');
											});	
											next();
										}
									], function() {
										console.log('successfully created nas folder.');
									});
								});
							}

							return;
						}
							
						// Mount all
						device.mount(function(err, mountPoint) {

							self.sataDevices.push({
								deviceName: device.deviceName,
								mountPoints: [ mountPoint ]
							});
							//console.log(device.deviceName + 'Mounted to ' + mountPoint);
						});
					});

				});
			}

			callback();
		});

		// Listen to event
		disk.on('BlockDeviceAdded', self.blockDeviceAddedHandler);
		disk.on('BlockDeviceRemoved', self.blockDeviceRemovedHandler);

	});
};

Disk.prototype.blockDeviceAddedHandler = function(deviceName) {
	var self = this;

	disk.getBlockDevice(deviceName, function(err, device) {
		if (err)
			return;

		device.getDrive(function(err, drive) {
			if (err)
				return;

			// Mount removable device automatically
			if (!drive.removable)
				return;
			//mountPoint = "/mnt";
			device.mount(function(err, mountPoint) {
				if (err)
					return;

				console.log('Mounted on ' + mountPoint);
			});
		});
	});
};

Disk.prototype.blockDeviceRemovedHandler = function(deviceName) {
	var self = this;
};

// Metadata
module.exports = {
	name: 'Disk',
	prototype: Disk
};
